### ISSUE TITLE
What exactly do the sgd parameters (batch_size, learning_momentum, valid_size, f_stable) and verbose= True output do/mean?

### ISSUE THREAD

**Issue by bridgesra** [aigamedev/scikit-neuralnetwork#117]
@alexjc Thanks for the helpful comments here: https://github.com/aigamedev/scikit-neuralnetwork/issues/104 !

I'm trying to get a NN configured using sknn.mlp's Regressor class. Basically, I'm trying to know exactly what the parameters of the class (specifically, valid_size, n_stable, f_stable, learning_momentum, and batch_size) do more specifically than I've found in the documentation. 

Here's what I think. Feel free to correct/ add/ destroy of just write explanations!

 Let's say batch_size = 10, valid_size = .25, and my training set has 100 points. To start an epoch it will take 75 of the data points and for each point compute the error between the predicted output and the actual.  Then it will use this to decide how to change each of the NN weights. It doesn't actually change any of the weights until it's finished w/ the 75, then changes each weight by the average over the 75 errors. Is that correct?

To finish the epoch, it checks the cost function on the held-out 25 points using the updated model's predictions. If this cost is less than f_stable, this epoch is tallied towards the n_stable quota. Is that correct?

When we have verbose = True, then the validation error that is printed is what? I guess it's the mean squared error on the 25 held out points for that epoch. Is that right?

Lastly, n_stable is a total count of the stable epochs, not the number of stable epochs that must occur consecutively; e.g.,  if num_stable = 3, then epochs 1,2, and 4 are stable it will end the training, right?

Can you tell me more concretely how learning_momentum changes the update step?

Less verbose questions related to this were posted on StackOverflow: 
http://stackoverflow.com/questions/32958847/what-exactly-does-valid-size-f-stable-n-stable-do-in-scikit-neuralnetwork-fitt

http://stackoverflow.com/questions/32957564/configuration-of-neural-network-for-regression-using-scikit-neuralnetwork

Thanks for any help!

**Comment by alexjc** [comment]
You have most of them correct!
`valid_size` is the percentage of the data to be held out for validation, 0.2 means 20%.
`verbose` is a flag that enables the displaying of error (as specified in constructor) on validation set.
`f_stable` is the tolerance of improvement on validation set, e.g. if below `0.001` it's considered stable.
`n_stable` is the number of consecutive iterations after which the training loop bails out.

For the rest:
`momentum` is a standard neural network learning parameter for stochastic gradient descent. It doesn't change the behavior of parameters above.
`batch_size` is also used exactly as standard for neural networks, it also doesn't affect any of the parameters above. (Directly passed to PyLearn2.)

I hope this helps!

**Comment by bridgesra** [comment]
When we have `verbose = True`, the middle column is `Validation Error`, so is that mean squared error on the validation set using the newly updated model? or something else? 

for `f_stable`, I'm still not sure what that means. After changing the parameters for the epoch, if the overall error on the validation set improved less than `f_stable`, then it counts as a stable epoch? What if it got worse, so the new model did worse on the the validation set and improvement is negative and certainly counted? 

Thanks for the reply!

**Comment by alexjc** [comment]
MSE is the default error, so unless you changed it, yes. If the new model is worse, then it's counted as stable because only the best model should be returned and that hasn't changed.
